File: 2021/09/27/main.py
class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        h = {}
    
        def clean(email):
            i = email.index('@')
            user, atDomain = email[:i], email[i:]
            return deleteDot(trim(user)) + atDomain
        
        def deleteDot(username):
            return username[::-1].replace(".", "")

        def trim(username):
            i = username.find('+')
            if i == -1:
                return username
            return username[:i]
            
        for email in emails:
            h[clean(email)] = True
            
        return len(h)

File: 2021/09/27/main.go
package main

import "strings"

func numUniqueEmails(emails []string) int {
	h := make(map[string]bool, 100)
	for _, email := range emails {
		h[clean(email)] = true
	}
	return len(h)
}

func clean(email string) string {
	i := strings.IndexByte(email, '@')
	user, atDomain := email[:i], email[i:]
	return deleteDot(trim(user)) + atDomain
}

func deleteDot(username string) string {
	return strings.Replace(username, ".", "", -1)
}

func trim(username string) string {
	i := strings.IndexByte(username, '+')
	if i == -1 {
		return username
	}
	return username[:i]
}
